dlt/sources/helpers/requests/retry.py

```diff
@@ -0,0 +1,160 @@
+from email.utils import parsedate_tz, mktime_tz
+from functools import partial
+import inspect
+import re
+import time
+from typing import Optional, cast, Callable, Type, Union, Sequence, Tuple, List
+
+from requests import Response, HTTPError, ConnectionError, Timeout, Session as BaseSession
+from tenacity import Retrying, retry_if_exception_type, stop_after_attempt, RetryCallState, retry_any, wait_exponential
+from tenacity.retry import retry_base
+
+from dlt.sources.helpers.requests.session import Session, DEFAULT_TIMEOUT, TSession
+from dlt.common.typing import TimedeltaSeconds
+
+
+DEFAULT_RETRY_STATUS = (429, *range(500, 600))
+DEFAULT_RETRY_EXCEPTIONS = (ConnectionError, Timeout)
+DEFAULT_RETRY_ATTEMPTS = 5
+
+RetryPredicate = Callable[[Optional[Response], Optional[BaseException]], bool]
+
+
+def _get_retry_response(retry_state: RetryCallState) -> Optional[Response]:
+    ex = retry_state.outcome.exception()
+    if ex:
+        if isinstance(ex, HTTPError):
+            return cast(Response, ex.response)
+        return None
+    result = retry_state.outcome.result()
+    return result if isinstance(result, Response) else None
+
+
+class retry_if_status(retry_base):
+    """Retry for given response status codes"""
+
+    def __init__(self, status_codes: Sequence[int]) -> None:
+        self.status_codes = set(status_codes)
+
+    def __call__(self, retry_state: RetryCallState) -> bool:
+        response = _get_retry_response(retry_state)
+        if response is None:
+            return False
+        result = response.status_code in self.status_codes
+        return result
+
+
+class retry_if_predicate(retry_base):
+    def __init__(self, predicate: RetryPredicate) -> None:
+        self.predicate = predicate
+
+    def __call__(self, retry_state: RetryCallState) -> bool:
+        response = _get_retry_response(retry_state)
+        exception = retry_state.outcome.exception()
+        return self.predicate(response, exception)
+
+
+class wait_exponential_retry_after(wait_exponential):
```

Comment:
this code looks good. why did you decide to write it?
1. we could probably re-use logic from `tenacity`?
2. why not to use `HTTPAdapter` which does exactly what you do here? https://majornetwork.net/2022/04/handling-retries-in-python-requests/
```
adapter = TimeoutHTTPAdapter(timeout=(3, 60), max_retries=Retry(total=5, backoff_factor=1.5, allowed_methods=False, status_forcelist=[429, 500, 502, 503, 504]))
```

Reply:
there may be good reason to write that from scratch, I'm curious why :)

Reply:
Yeah it's basically the same for the common case.    
But I'm thinking from the POV of having to deal with wonky APIs that are not very standards compliant, or just have a couple of edge cases where the status code is ambigious. I usually need to add at least a little bit of custom logic to decide whether to retry so being able to inject some callback to inspect the response is super handy.  
E.g. think cases like where `400` code is used both for rate limit and "invalid input" kind of error. Or the extreme case, I work with one internal API which returns 200 on everything and error code is in the body.   

There is a lot of flexibility in tenacity that could be exposed as well for more exotic use cases. E.g. dynamic delays (say one error takes longer to recover than others).